### PR TITLE
Encode 'task_args' and 'task_kwargs' in 'store_result'.

### PR DESCRIPTION
**Pull request by arnau126**
Encode `task_args` and `task_kwargs` in `DatabaseBackend.store_result()` (as we do with `meta` and `result`) so we can retrieve them as python objects later on.

Note that this PR changes the serialization format of these fields.
For example, if `args=['a', 1, True, None]`
Currently it's stored as :
    `"['a', 1, True, None]"`
and after this PR it will be stored as:
    `'["a", 1, true, null]'` (json format)

### CODE REVIEW

**Comment by auvipy** [comment]
Is this a backward incompatible change?

**Comment by arnau126** [comment]
I think that this change won't break any main functionality, as these fields are just informative or for debugging.
As I said the difference is the way these fields are stored in the DB:
    `"['a', 1, True, None]"` vs. `'["a", 1, true, null]''`

For me it's not a problem that the storing format change.

Maybe it's a problem for someone who has a script or something that parse and use these fields. However, these fields are quiet new (the deploy was 2 days ago).

**Comment by kiyoya** [comment]
Could we make it an option?

Yes, the new release was landed in a few days ago. However the feature was merged several months ago (#36) and not a few people have been using `HEAD` instead of 1.0.1 (#61).

**Comment by arnau126** [comment]
I don't like the idea to make it an option. I think that not encoding these fields is a bug, because serializing a `list` or a `dict` using `str()` force you to use `eval()` to retrieve them as python objects, which is not safe.

**Comment by kiyoya** [comment]
It's a bit off-topic but there is `ast.literal_eval` for parsing strings like `task_args`.

Yes, I agree that`task_args` and `task_kwargs` should have been implemented like this in the first place. Unfortunately they weren't, however, and I'm not sure if it is worth making an incompatible change while there is a safe way to parse them.

Another possibility is to introduce new fields in parallel.

File: django_celery_results/backends/database.py
Comment (liquidpele) [inline django_celery_results/backends/database.py]: The args and kwargs encoding shouldn't match the result encoding, which is what this logic essentially does since you pass in obj and decode_content checks obj.content_encoding.  You ignore the returned content_encoding for those items above, so it's basically lost information.   Result encoding can be binary for instance, but args/kwargs may not be, and vice versa.

File: t/unit/backends/test_database.py
Comment (liquidpele) [inline t/unit/backends/test_database.py]: Are these the only types of inputs that can be received?  What about other things?  

You really need to look at all the different things that the kombu dumps function (which is what actually handles...  

https://github.com/celery/kombu/blob/master/kombu/serialization.py#L161

File: django_celery_results/backends/database.py
Comment (arnau126) [reply django_celery_results/backends/database.py]: I use `self.decode_content(...)` to deserialize because I used `self.encode_content(...)` to serialize. I thought that it was consistent. I don't know any other way to do it.

File: t/unit/backends/test_database.py
Comment (arnau126) [reply t/unit/backends/test_database.py]: I think that `args` is always a list and `kwargs` always a dict.
So, you mean something like:
```
request.args = ['a', 1, ['2', 3, {'b': 'x'}], {'c': 'd', 'e': 3, 'f': [5, 6, 7]}]
```
?

Comment (liquidpele) [reply t/unit/backends/test_database.py]: Yes, but an array/dict containing what?

File: django_celery_results/backends/database.py
Comment (liquidpele) [reply django_celery_results/backends/database.py]: Look at how the decode works.

Comment (arnau126) [reply django_celery_results/backends/database.py]: Look at how encode works.

File: t/unit/backends/test_database.py
Comment (arnau126) [reply t/unit/backends/test_database.py]: The `content_encoding` that returns the `encode_content` function only depends on `self.serializer`. Which means that it will be the same for `result`, `meta`, `task_args`, `task_kwargs` and for all others objects serialized using the `encode_content` of the same DatabaseBackend instance.

Moreover, we're also ignoring the `content_encoding` of `meta` and there is error when we deserialize it using the same `decode_content` function.

File: django_celery_results/backends/database.py
Comment (arnau126) [reply django_celery_results/backends/database.py]: The `content_encoding` that returns the `encode_content` function only depends on `self.serializer`. Which means that it will be the same for `result`, `meta`, `task_args`, `task_kwargs` and for all others objects serialized using the `encode_content` of the same DatabaseBackend instance.

Moreover, we're also ignoring the `content_encoding` of `meta` and there is error when we deserialize it using the same `decode_content` function.

File: t/unit/backends/test_database.py
Comment (arnau126) [reply t/unit/backends/test_database.py]: I don't get you.
Is it okay or not this example: `request.args = ['a', 1, ['2', 3, {'b': 'x'}], {'c': 'd', 'e': 3, 'f': [5, 6, 7]}]` ?
Let me know which other data types I could include.

**Comment by arnau126** [comment]
Rebased

**Comment by auvipy** [comment]
> Rebased

let's reach a consensus for the changes before the merge.

**Comment by arnau126** [comment]
I'm pretty sure that the code works.
The important issue here is the backward incompatibility, which @kiyoya and I already discussed.
I'm willing to hear other points of view.

**Comment by auvipy** [comment]
Keeping BC is very important.

**Comment by arnau126** [comment]
So the options are:
- Make it an option (maybe a setting).
- Introduce new fields in parallel.
- Wait until a major release where backward incompatibility are allowed.
- Don't do it at all (reject the PR).

Let me know if you come up with other ideas.

File: t/unit/backends/test_database.py
Comment (liquidpele) [reply t/unit/backends/test_database.py]: What I mean, is what if I pass an object as an arg?  What about somethign that's not json serializable like django's lazy gettext objects?  What about a file handle?  A string that's 1 million bytes?  etc.  Your tests handle the easy test cases, but those are the kinds of tests that aren't particularly helpful... try to break your own code ;)

File: django_celery_results/backends/database.py
Comment (liquidpele) [reply django_celery_results/backends/database.py]: obj.content_encoding is set according to result, but encoding doesn't take that into account, it only looks at the data for the field in question.  What you've done is made a case where the result may be binary but it encoded your args as json... but then trying to decode the args would try to decode the binary which it wasn't encoded as.  

Basically content_encoding was only ever meant for result.  If you want to encode the args/kwargs you need to make sure it always handles them the same way, otherwise there will be a lot of inconsistency.

Comment (arnau126) [reply django_celery_results/backends/database.py]: > obj.content_encoding is set according to result

False, it is not set according to `result`. It is set according to `self.serializer`, which is always the same.

Kombu `dumps()` function would only take `result` into account if `self.serializer` were `None` or `"raw"`. But this is not possible because celery's allowed serializers are only: json, pickle, yaml and msgpack (or custom serializer, but not None nor "raw").

Comment (liquidpele) [reply django_celery_results/backends/database.py]: Hmm, maybe I'm reading this wrong.  What is the obj args in decode_content?  I was assuming that was the django model instance.

Comment (arnau126) [reply django_celery_results/backends/database.py]: I added more test where:
- result is a dict
- result is text
- result is bytes

to ensure that it works fine with all of them.

File: t/unit/backends/test_database.py
Comment (arnau126) [reply t/unit/backends/test_database.py]: I included the custom testing class `SomeClass` in args and kwargs to make sure that it is able to serialize non standard objects.

Comment (liquidpele) [reply t/unit/backends/test_database.py]: awesome thanks!

File: django_celery_results/backends/database.py
Comment (arnau126) [reply django_celery_results/backends/database.py]: > Hmm, maybe I'm reading this wrong. What is the obj args in decode_content? I was assuming that was the django model instance.

The `obj` argument of the `decode_content()` function is a `TaskResult` instance.

Comment (liquidpele) [reply django_celery_results/backends/database.py]: ah, okay, then this makes much more sense, sorry about the confusion.  If all your extra tests are passing then I'm good besides @auvipy 's concern about backward compatibility.

**Comment by auvipy** [comment]
> So the options are:
> 
> * Make it an option (maybe a setting).
> * Introduce new fields in parallel.
> * Wait until a major release where backward incompatibilities are allowed.
> * Don't do it at all (reject the PR).
> 
> Let me know if you come up with other ideas.

I would wait for celery 5 and this package 2.0 version

**Comment by KevinPoole** [comment]
> It's a bit off-topic but there is `ast.literal_eval` for parsing strings like `task_args`.
> 
> Yes, I agree that`task_args` and `task_kwargs` should have been implemented like this in the first place. Unfortunately they weren't, however, and I'm not sure if it is worth making an incompatible change while there is a safe way to parse them.
> 
> Another possibility is to introduce new fields in parallel.

Looks like this particular PR has already been settled as "not yet" but I just wanted to comment on the above: `ast.literal_eval` is insufficient for cases where, for example, `pickle` has been used for serialization and there exists some `CustomClass` objects in the kwargs. In this case `ast.literal_eval` chokes `invalid syntax` when you get to the `repr` string for the objects.

**Comment by ZviBaratz** [comment]
Is there anything I can do to help move this forward?

**Comment by arnau126** [comment]
Hope that now that Celery 5 is released, it's a good moment to update this package to version 2.0, and merge this PR.

**Comment by auvipy** [comment]
is it possible to not to break existing systems?

**Comment by arnau126** [comment]
This PR will only change how new TaskResult instances will store `task_args` and `task_kwargs`. Already existing instances will keep the current format.
So the problem is that we will have old instances with these fields serialized using `str()` and new instances with these fields serialized using `json.dumps()`

I already suggested a few options:
> - Make it an option (maybe a setting).
> - Introduce new fields in parallel.
> - **Wait until a major release where backward incompatibilities are allowed.**
> - Don't do it at all (reject the PR).

However, I think that store a `list` or a `dict` using `str()` is not a good practice, so it shouldn't be an option at all.
That's why I would wait until a major release.
And I think that now is a good moment to do it: celery 5 is released, py35 support has been dropped...

**Comment by johnthealy3** [comment]
I just encountered this after solving a similar problem trying to use the JSONField (I was using the postgres contrib one but now merged in Django 3.1: https://docs.djangoproject.com/en/3.1/ref/contrib/postgres/fields/#jsonfield) for task_args and task_kwargs. Perhaps now is the time, with an option or with Django version checking?

Happy to provide some code if this approach has potential.

**Comment by ZviBaratz** [comment]
@arnau126 I tried to check out this PR and I'm still getting `task_args` and `task_kwargs` encoded as strings. Any idea what I'm missing?

**Comment by arnau126** [comment]
`task_args` and `task_kwargs` still are `TextField` fields, so when you retrieve a TaskResult instance from DB they are strings. To cast them to `list` or `dict` you have to do `json.loads`

**Comment by ZviBaratz** [comment]
@arnau126 that makes sense, sorry :sweat_smile:

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
I am accepting this PR, but new fields / better solutions using django new fields etc is wellcome

**Comment by robvdl** [comment]
So 2.0.0 didn't fix the CVE yet? issue #154 when I saw a new major release I had my hopes up the CVE was fixed but I don't think it has yet. Has version 2.0.0 been flagged yet or is it a matter of time that version gets flagged too?

**Comment by arnau126** [comment]
@robvdl , this PR wasn't meant to fix #154 at all, but to change the way the input arguments were stored (use `json.dumps` instead of `str`).

**Comment by igorvoltaic** [comment]
> Encode `task_args` and `task_kwargs` in `DatabaseBackend.store_result()` (as we do with `meta` and `result`) so we can retrieve them as python objects later on.
> 
> Note that this PR changes the serialization format of these fields.
> For example, if `args=['a', 1, True, None]`
> Currently it's stored as :
> `"['a', 1, True, None]"`
> and after this PR it will be stored as:
> `'["a", 1, true, null]'` (json format)

Seems like not really working. This is what i need to do to get back the task arguments.
```
>>> from django_celery_results.models import TaskResult
>>> import ast
>>> import json
>>> TaskResult.objects.last()
<TaskResult: <Task: c4e59e69-d324-44fd-95ad-1a8c210fa2bd (SUCCESS)>>
>>> t = TaskResult.objects.last()
>>> t.task_args
'"({\'dataset_id\': 2, \'height\': 2600, \'width\': 2600, \'plot_type\': \'bar\', \'params\': {\'x\': \'month\', \'y\': \'passengers\', \'hue\': None}, \'columns\': [\'month\', \'passengers\'], \'id\': 1},)"'
>>> json.loads(t.task_args)
"({'dataset_id': 2, 'height': 2600, 'width': 2600, 'plot_type': 'bar', 'params': {'x': 'month', 'y': 'passengers', 'hue': None}, 'columns': ['month', 'passengers'], 'id': 1},)"
>>> ast.literal_eval(json.loads(t.task_args))
({'dataset_id': 2, 'height': 2600, 'width': 2600, 'plot_type': 'bar', 'params': {'x': 'month', 'y': 'passengers', 'hue': None}, 'columns': ['month', 'passengers'], 'id': 1},)

>>> ast.literal_eval(json.loads(t.task_args))[0]
{'dataset_id': 2, 'height': 2600, 'width': 2600, 'plot_type': 'bar', 'params': {'x': 'month', 'y': 'passengers', 'hue': None}, 'columns': ['month', 'passengers'], 'id': 1}
```

Do I correctly understand this commit was aimed to allow simply do `json.loads(t.task_args)` ?

**Comment by auvipy** [comment]
this is a more recent PR https://github.com/celery/django-celery-results/pull/178 and @arnau126 opened a new PR in celery as well https://github.com/celery/celery/pull/6595

**Comment by arnau126** [comment]
> Do I correctly understand this commit was aimed to allow simply do json.loads(t.task_args) ?

Yes, you understand correctly.


> Seems like not really working

You're right, unfortunately not working, because of the new Celery [task_protocol=2](https://docs.celeryproject.org/en/stable/userguide/configuration.html#std-setting-task_protocol) and the [hide-sensitive-information feature](https://docs.celeryproject.org/en/stable/userguide/tasks.html#hiding-sensitive-information-in-arguments)

I have a work-in-progress PR (celery/celery#6595) to fix it (to allow simply do `json.loads(t.task_args)`).

**Pull request merged**
